src/actions/groups.cr

```diff
@@ -101,50 +105,72 @@ module DppmRestApi::Actions::Groups
-        route.query_parameters[key] = values - specified
+      if context.request.query_params.empty?
+        route.query_parameters.clear
+        group.permissions[path] = route
+        next group
+      end
+      route.query_parameters.each do |key, values|
+        next unless context.request.query_params.has_key? key
+        specified = context.request.query_params.fetch_all(key).reject &.empty?
+        if specified.empty?
+          route.query_parameters.delete key
+        else
+          route.query_parameters[key] = values - specified
+        end
       end
+      group.permissions[path] = route
+      group
     end
-    group.permissions[path] = route
-    DppmRestApi.permissions_config.groups[group_idx] = group
+    raise NoSuchGroup.new context, group_id_s unless group_already_found
     DppmRestApi.permissions_config.sync_to_disk
   end
+
   relative_delete "/:id/route/:path" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
     group_id = (group_id_s = context.params.url["id"]).to_i?
     raise BadRequest.new context,
       %[Group ID "#{group_id_s}" was not an integer] if group_id.nil?
-    group_idx = DppmRestApi.permissions_config.groups.index { |grp| grp.id == group_id }
-    raise NoSuchGroup.new context, group_id_s if group_idx.nil?
-    group = DppmRestApi.permissions_config.groups[group_idx]
-    deleted = group.permissions.delete context.params.url["path"]
-    raise NotFound.new context,
-      "no permissions found at the specified path (possibly already \
-        deleted?)" if deleted.nil?
+    group_already_found = false
+    DppmRestApi.permissions_config.groups.map! do |group|
+      next group if group.id != group_id
+      raise DuplicateGroup.new context, group_id_s if group_already_found
+      group_already_found = true
+      group.permissions.delete(
+        context.params.url["path"]
+      ) || raise NotFound.new context, "\
+                              no permissions found at the specified path \
+                              (possibly already deleted?)"
+      group
+    end
+    raise NoSuchGroup.new context, group_id_s unless group_already_found
     DppmRestApi.permissions_config.sync_to_disk
   end
+
   relative_delete "/:id" do |context|
     raise Unauthorized.new context unless Actions.has_access? context, Access::Update
     group_id = (group_id_s = context.params.url["id"]).to_i?
     raise BadRequest.new context,
       %[Group ID "#{group_id_s}" was not an integer] if group_id.nil?
-    to_be_deleted = DppmRestApi.permissions_config
+    group_already_found = false
+    DppmRestApi.permissions_config
       .groups
-      .index { |group| group.id == group_id }
-    raise NoSuchGroup.new context, group_id_s if to_be_deleted.nil?
-    DppmRestApi.permissions_config.groups.delete_at to_be_deleted
+      .reject! do |group|
+        if group.id == group_id
+          raise DuplicateGroup.new context, group_id_s if group_already_found
```

Comment:
I don't think it's good.
What does it mean for the user if (s)he gets this error, (s)he has to execute the request again?
This check of duplicated ID should be done at the group creation, only.